drivers/volume/portworx/portworx.go

```diff
@@ -923,6 +923,140 @@ func (p *portworx) SnapshotDelete(snapDataSrc *crdv1.VolumeSnapshotDataSource, _
 	}
 }
 
+// VolumeSnapshotRestore does in-place restore of snapshot to it's parent volume
+func (p *portworx) VolumeSnapshotRestore(snapRestore *stork_crd.SnapshotRestore) error {
+	var err error
+	// restoreVolumes[snapID]volumeID
+	restoreVolumes := make(map[string]string)
+	// should bw used to set security context if needed
+	params := make(map[string]string)
+	snapName := snapRestore.Spec.SourceName
+	snapNamespace := snapRestore.Spec.SourceNamespace
+
+	// TODO: Restoring snapshot to volume other than parent volume is not supported by PX
+	if snapRestore.Spec.DestinationPVC != nil {
+		return fmt.Errorf("restore to volume other than parent is not supported")
+	}
+	if snapRestore.Spec.GroupSnapshot {
+		logrus.Infof("GroupVolumeSnapshot In-place restore request for %v", snapName)
+		// restore each snapshot to their own parents
+		restoreVolumes, err = getRestoreVolumeMap(snapName, snapNamespace)
+		if err != nil {
+			return err
+		}
+	} else {
+		logrus.Infof("VolumeSnapshot In-place restore request for %v", snapName)
+		// GetSnapshot Details
+		snapshot, err := k8s.Instance().GetSnapshot(snapName, snapNamespace)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve VolumeSnapshot %s: %v",
+				snapName, err)
+		}
+		logrus.Infof("Snapshot found %v", snapshot.Metadata.Name)
+		if len(snapshot.Spec.SnapshotDataName) == 0 {
+			return fmt.Errorf("VolumeSnapshot %s is not bound to any VolumeSnapshotData",
+				snapName)
+		}
+
+		snapshotData, err := k8s.Instance().GetSnapshotData(snapshot.Spec.SnapshotDataName)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve VolumeSnapshotData %s: %v",
+				snapshot.Spec.SnapshotDataName, err)
+		}
+		// restore to snapshot PVC
+		volID, err := getVolumeIDFromPVC(
+			snapshot.Spec.PersistentVolumeClaimName,
+			snapshot.Metadata.Namespace)
+		if err != nil {
+			return fmt.Errorf("failed to get volume ID for snapshot %v", err)
+		}
+
+		snapID := snapshotData.Spec.PortworxSnapshot.SnapshotID
+		restoreVolumes[snapID] = volID
+	}
+	logrus.Debugf("Restore volume map %v", restoreVolumes)
+
+	switch snapRestore.Spec.SourceType {
+	case "", string(crdv1.PortworxSnapshotTypeLocal):
+		return p.localSnapshotRestore(restoreVolumes, params)
+	case string(crdv1.PortworxSnapshotTypeCloud):
+		return p.cloudSnapshotRestore(restoreVolumes, params)
```

Comment:
Return error for cloudsnap since that isn't supported right now. And we'll need additional changes here for that.

Reply:
I see that `cloudSnapshotRestore` is returning an error